// node-dependency-injection/dist/bin/Services/File/BaseAdapter.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

var BaseAdapter = /*#__PURE__*/function () {
  function BaseAdapter() {
    _classCallCheck(this, BaseAdapter);
  }

  _createClass(BaseAdapter, [{
    key: "defaultConfiguration",
    get:
    /**
     * @return {object}
     */
    function get() {
      return {
        parameters: {
          'some.parameter.foo': 'foo',
          'some.parameter.bar': 'bar'
        },
        imports: [{
          resource: "/path/to/another-file.".concat(this.constructor.FORMAT)
        }],
        services: {
          'some.service.key': {
            "class": './some/class/path',
            arguments: ['@some.service.key2', '%some.parameter.foo%']
          },
          'some.service.key2': {
            "class": './some/class/path2',
            arguments: ['%some.parameter.bar%']
          }
        }
      };
    }
  }]);

  return BaseAdapter;
}();

exports["default"] = BaseAdapter;
